### DISCUSSION TITLE
Fitting

### DISCUSSION THREAD

**Discussion by atravert** [spectrochempy/spectrochempy#393] [Developer discussions]
I am starting using the fitting module in practical cases, and I take this opportunity to jote down few notes/questions about its current implementation. There is no need of immediate response as there will be probably more points and editing as I will =use  it :-)

- 1. there is currently a small error in the model script used in  [userguide/analysis/fitting](https://www.spectrochempy.fr/latest/userguide/analysis/fitting.html#NDDataset-modelling-using-the-Fit-method): the actual peak is outside the limits of the first line ! so the limits should be set to e.g.  3600.0 - 3700.0, to get a correct fit :-) 

- 2. there is also probably an error line 77 of models.py: print(newargs) should be removed or commented to avoid unwanted output (?)

- 3. no linear baseline should be automatically calculated and added, in particular because in some instances, the user will already have a "baseline corrected" dataset. This is the case of the example in [userguide/analysis/fitting](https://www.spectrochempy.fr/latest/userguide/analysis/fitting.html#NDDataset-modelling-using-the-Fit-method). Here the spectrum is baseline corrected and in this case the baseline add 2 degrees of freedom that are not needed, and increase the number of iterations (+22% in this case). In more complicated cases, it can be worse (observed up to +70%), and also worsen the fit. Scpy has very good and sophisticated baseline correction schemes, they can be are somehow 'wasted' with this automatic linear correction.   

- 4. the amplitude setting is not straightforward to adjust manually because (1) it corresponds to the area of the peak, and (2) it is renormalized when runing with dry=True, and this renormalization depends on the presence of other peaks.... In practice i think it would be easier for the user to set the height  (because it is easier to estimate), than the area, and these heights should no be renormalized when dry=True.

- 5. the `silent` parameter should be passed to the scipy.optimize functions (currently `(..., disp=False, ...)`, we could pass a `disp` parameter to `fitting.optimize()` and call it with `disp = not silent` . By the way we should use a consistent keyword (and default) in scpy.analysis functions (mcrals: verbose=False ;  nnmf: verbose=True).      

- 6. maybe not necessary to generate a .model attribute to the dataset when Fit.run() is used. Instead, Fit/run() could return the model lines in a NDDataset on its own ?   it would allow to have name of each line thet Coord labls, and properties of each line in desfription, etc... Ploting of the model and experimental dataset could be done with standard scpy plotting functions instead using ("plot_model=True") ? Another possibility would be to add this dataset as an attribute of the fitting object and to add methods consitent with other analysis methods -- i.e. reconstruct() and plotmerit() ?

**Comment by fernandezc** [comment]
You are certainly right about most things. I must admit that I did not work much on this module that I had previously written for a program dealing with NMR. 
Concerning the automatic baseline, I have a little trouble to see why it is a problem when the baseline is already (well) corrected. But I didn't think much about your remark. 
For the amplitude, it's true that it can be a problem, to have to estimate the area, but the advantage is that it returns directly the concentrations? But there is certainly other way to do it. 
For the other points and those mentioned above, I let you propose modifications, no worries!
